File: python/copass-core/src/copass_core/context_window.py
"""Context Window — agent conversation as an ephemeral data source.

Port of ``typescript/packages/core/src/context-window/`` — combines
``context-window.ts``, ``resource.ts``, and ``types.ts`` into a single
module (Python doesn't need the TS split).

Each :meth:`ContextWindow.add_turn` call appends to a local buffer
AND pushes through the underlying ``DataSource`` so the thread itself
becomes retrievable. Pass the window to any ``client.retrieval.*``
call to get window-aware retrieval without hand-managing ``history``.

Construct via :meth:`ContextWindowResource.create` for a fresh
thread or :meth:`ContextWindowResource.attach` to resume an existing
one.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Any, Dict, List, Optional

from copass_core.data_sources import BaseDataSource
from copass_core.types import ChatMessage

if TYPE_CHECKING:
    from copass_core.client import CopassClient


def _capitalize_role(role: str) -> str:
    if not role:
        return role
    return role[0].upper() + role[1:]


class ContextWindow(BaseDataSource):
    """An agent's conversation wrapped as an ephemeral data source.

    Satisfies ``WindowLike`` (``get_turns()``) so the
    ``RetrievalResource`` accepts it directly as a ``window=`` arg.
    """

    def __init__(
        self,
        *,
        client: "CopassClient",
        sandbox_id: str,
        data_source_id: str,
        project_id: Optional[str] = None,
        initial_turns: Optional[List[ChatMessage]] = None,
    ) -> None:
        super().__init__(
            client=client,
            sandbox_id=sandbox_id,
            data_source_id=data_source_id,
            project_id=project_id,
        )
        self._turns: List[ChatMessage] = list(initial_turns or [])

    async def add_turn(self, turn: ChatMessage) -> None:
        """Append a turn and push it through the underlying data
        source.

        Awaits the push so ingestion failures surface at the call
        site. Callers wanting fire-and-forget can wrap in
        ``asyncio.create_task``.

        The turn's ``name`` (if set) is forwarded as the envelope's
        ``speaker`` field; absent ``name`` falls back to the
        capitalized ``role`` (``"user"`` → ``"User"``,
        ``"assistant"`` → ``"Assistant"``). The
        ``f"{turn.role}: {turn.content}"`` content-prefix munging the
        prior version used has been retired — the wire body is now
        the message ``content`` verbatim, with ``speaker`` riding on
        the envelope.
        """
        self._turns.append(turn)
        speaker = turn.name if turn.name else _capitalize_role(turn.role)
        await self.push(
            turn.content,
            source_type="conversation",
            speaker=speaker,
        )

    def get_turns(self) -> List[ChatMessage]:
        """Defensive copy of the turn log — callers can't mutate
        internal state."""
        return list(self._turns)

    async def close(self) -> None:
        """Mark the underlying source as disconnected. Idempotent on
        the server."""
        await self.disconnect()


class ContextWindowResource:
    """Factory for :class:`ContextWindow` instances.

    Accessed via ``client.context_window``. :meth:`create` registers
    a new ephemeral data source and returns a window bound to it;
    :meth:`attach` rehydrates a window against an existing
    ``data_source_id``.
    """

    def __init__(self, client: "CopassClient") -> None:
        self._client = client

    async def create(
        self,
        *,
        sandbox_id: str,
        project_id: Optional[str] = None,
        name: Optional[str] = None,
    ) -> ContextWindow:
        """Register a fresh ephemeral data source and return a window
        bound to it."""
        source_name = name if name is not None else f"window-{int(time.time() * 1000)}"
        source = await self._client.sources.register(
            sandbox_id,
            provider="custom",
            name=source_name,
            ingestion_mode="manual",
            kind="ephemeral",
        )
        return ContextWindow(
            client=self._client,
            sandbox_id=sandbox_id,
            data_source_id=source["data_source_id"],
            project_id=project_id,
        )

    async def attach(
        self,
        *,
        sandbox_id: str,
        data_source_id: str,
        project_id: Optional[str] = None,
        initial_turns: Optional[List[ChatMessage]] = None,
    ) -> ContextWindow:
        """Reattach to an existing source — typically one the caller
        persisted after an earlier :meth:`create`.
        """
        source = await self._client.sources.retrieve(sandbox_id, data_source_id)
        return ContextWindow(
            client=self._client,
            sandbox_id=sandbox_id,
            data_source_id=source["data_source_id"],
            project_id=project_id,
            initial_turns=initial_turns,
        )


__all__ = ["ContextWindow", "ContextWindowResource"]

File: typescript/packages/core/src/context-window/context-window.ts
import { BaseDataSource, type BaseDataSourceOptions } from '../data-sources/base.js';
import type { ChatMessage } from '../resources/retrieval.js';

export interface ContextWindowOptions extends BaseDataSourceOptions {
  /** Pre-existing turns to seed the local buffer, e.g. when resuming. */
  initialTurns?: ChatMessage[];
  /**
   * Default participant roster for this conversation. Forwarded as the
   * envelope's `participants` field on every turn pushed through
   * {@link addTurn} unless that call passes its own override. Useful
   * for downstream pronoun resolution — when the LLM sees
   * `[Speaker: User]` and `[Participants: User, Assistant]` it can
   * resolve "your X" against the other listed participant. Per-call
   * overrides win over the constructor default.
   */
  participants?: string[];
}

/**
 * An agent's conversation wrapped as an ephemeral data source.
 *
 * Each call to {@link addTurn} appends a turn to the local buffer and pushes
 * its content through the underlying data source, so the thread itself
 * becomes retrievable. Pass the window to any `client.retrieval.*` call to
 * get window-aware retrieval without hand-managing a `history` array.
 *
 * Construct via {@link ContextWindowResource.create} for a fresh thread or
 * {@link ContextWindowResource.attach} to resume an existing one.
 */
export interface AddTurnOptions {
  /**
   * Per-call participants override. Defaults to the window's
   * constructor-time `participants`, if any.
   */
  participants?: string[];
}

export class ContextWindow extends BaseDataSource {
  private readonly turns: ChatMessage[];
  private readonly participants?: readonly string[];

  constructor(options: ContextWindowOptions) {
    super(options);
    this.turns = [...(options.initialTurns ?? [])];
    this.participants = options.participants
      ? Object.freeze([...options.participants])
      : undefined;
  }

  /**
   * Append a turn and push it through the underlying data source.
   *
   * Awaits the push so ingestion failures surface at the call site. Callers
   * wanting fire-and-forget can drop the `await` or wrap in `void`.
   *
   * The turn's `name` (if set) is forwarded as the envelope's
   * `speaker` field; absent `name` falls back to the capitalized
   * `role` (`'user'` → `'User'`, `'assistant'` → `'Assistant'`). The
   * `${role}: ${content}` content-prefix munging the prior version
   * used has been retired — the wire body is now the message
   * `content` verbatim, with `speaker` riding on the envelope.
<<<<<<< HEAD
   *
   * Participants come from the call-site override if set, otherwise
   * from the window's constructor-time roster, otherwise omitted.
=======
>>>>>>> 1d8137ce5ea4a299db96d55a116303f124248bb6
   */
  async addTurn(turn: ChatMessage, options: AddTurnOptions = {}): Promise<void> {
    this.turns.push(turn);
    const speaker = turn.name ?? capitalizeRole(turn.role);
<<<<<<< HEAD
    const participants =
      options.participants ??
      (this.participants ? [...this.participants] : undefined);
    await this.push(turn.content, {
      sourceType: 'conversation',
      speaker,
      participants,
=======
    await this.push(turn.content, {
      sourceType: 'conversation',
      speaker,
>>>>>>> 1d8137ce5ea4a299db96d55a116303f124248bb6
    });
  }

  /** Current turn log — returned as a defensive copy so callers can't mutate internal state. */
  getTurns(): ChatMessage[] {
    return [...this.turns];
  }

  /** Mark the underlying source as disconnected. Best-effort — idempotent on the server. */
  async close(): Promise<void> {
    await this.disconnect();
  }
}

function capitalizeRole(role: string): string {
  if (!role) return role;
  return role[0]!.toUpperCase() + role.slice(1);
}
